typescript/schedules.md

```diff
@@ -0,0 +1,1246 @@
+  nextActionTimes: Date[];
+
+  createdAt: Date;
+  lastUpdatedAt: Date;
+
+  isValid(): boolean;
+
+  /** Error for invalid schedule. If this is present, no actions will be taken. */
+  invalidScheduleError?: string;
+};
+
+export interface ScheduleAction {
+  /** Time that the Action was scheduled for, including jitter. */
+  scheduledAt: Date;
+
+  /** Time that the Action was actually taken. */
+  takenAt: Date;
+
+  /** If action was {@link StartWorkflowAction}. */
+  workflow?: WorkflowExecution;
+  // TODO or WorkflowHandle? would be more convenient, eg `const latestResult = await recentActions.pop().result()`
+}
+
+export interface ScheduleClientOptions {
+  /**
+   * {@link DataConverter} to use for serializing and deserializing payloads
+   */
+  dataConverter?: DataConverter;
+
+  /**
+   * Used to override and extend default Connection functionality
+   *
+   * Useful for injecting auth headers and tracing Workflow executions
+   */
+  interceptors?: ScheduleClientInterceptors;
+
+  /**
+   * Identity to report to the server
+   *
+   * @default `${process.pid}@${os.hostname()}`
+   */
+  identity?: string;
+
+  /**
+   * Connection to use to communicate with the server.
+   *
+   * By default `ScheduleClient` connects to localhost.
+   *
+   * Connections are expensive to construct and should be reused.
+   */
+  connection?: ConnectionLike;
+
+  /**
+   * Server namespace
+   *
+   * @default default
+   */
+  namespace?: string;
+}
+
+export type ScheduleClientOptionsWithDefaults = Replace<
+  Required<ScheduleClientOptions>,
+  {
+    connection?: ConnectionLike;
+  }
+>;
+export type LoadedScheduleClientOptions = ScheduleClientOptionsWithDefaults & {
+  loadedDataConverter: LoadedDataConverter;
+};
+
+export function defaultScheduleClientOptions(): ScheduleClientOptionsWithDefaults {
+  return {
+    dataConverter: {},
+    // The equivalent in Java is ManagementFactory.getRuntimeMXBean().getName()
+    identity: `${process.pid}@${os.hostname()}`,
+    interceptors: {},
+    namespace: 'default',
+  };
+}
+
+// TODO
+// function ensureArgs<W extends Workflow, T extends ScheduleOptions<W>>(opts: T): Omit<T, 'args'> & { args: unknown[] } {
+//   const { args, ...rest } = opts;
+//   return { args: args ?? [], ...rest };
+// }
+
+interface SchdeduleHandleOptions extends GetSchdeduleHandleOptions {
+  workflowId: string;
+  runId?: string;
+  interceptors: ScheduleClientCallsInterceptor[];
+  /**
+   * A runId to use for getting the workflow's result.
+   *
+   * - When creating a handle using `getHandle`, uses the provided runId or firstExecutionRunId
+   * - When creating a handle using `start`, uses the returned runId (first in the chain)
+   * - When creating a handle using `signalWithStart`, uses the the returned runId
+   */
+  runIdForResult?: string;
+}
+
+/**
+ * Policy for overlapping Actions.
+ */
+export enum ScheduleOverlapPolicy {
+  /**
+   * Don't start a new Action.
+   */
+  Skip = 1,
```

Comment:
can these get imported/generated from the protos directly? I don't like duplicating enum values

Reply:
They can, but I don't like it from DX perspective: includes unspecified and is long 

```ts
import { temporal } from protos
temporal.api.enums.v1.schedule.ScheduleOverlapPolicy.SCHEDULE_OVERLAP_POLICY_UNSPECIFIED

// vs

import { ScheduleOverlapPolicy } from client
ScheduleOverlapPolicy.Skip
```

I can add this to ensure it stays in sync:

`checkExtends<temporal.api.enums.v1.schedule.ScheduleOverlapPolicy, ScheduleOverlapPolicy>();`

Reply:
haha, very true. well, then it's a question of inter-language consistency vs short names. I'll defer to sdk team.

Reply:
checkExtends would lead to compile time failure but you write a version of it that will do the validation.

I'd just use the proto definitions as we've done in the rest of the SDK for consistency.

You can re-export the enum from the client package.

Reply:
Here are all the cases of TS defining our own enums that match proto enums:

![image](https://user-images.githubusercontent.com/251288/185174432-cecb32b9-8dee-4c8b-9d28-64b20d99ca72.png)

I will match the all-caps convention used in other enums (I used the SentenceCase TS convention) and add UNSPECIFIED, and remove it if this is implemented before we leave beta: https://github.com/temporalio/temporal/issues/3240

Reply:
The reason we redefine these enums is because they're used in workflow code and enums in TS are more than types, importing the proto implementation would significantly increase the workflow bundle size.

Not sure if TS should keep unspecified or not, I would say no because we have optional types which makes UNSPECIFIED redundant (and confusing?) and yes because it's consistent with the rest of the API.

Reply:
I think that's a good reason to redefine enums, and I think DX is another good reason to.

If UNSPECIFIED was here to stay (hoping it's not), then I'm inclined to include in describe type but not in `ScheduleOptions` type. I was thinking we might need in describe type because if we translated UNDEFINED to SKIP, that might be inaccurate in future? (Server could change default, and then that schedule would behave differently than SKIP. We could change how we translate UNDEFINED in sync, but if user saves the pre-server-change description with SKIP in it, and referenced that, it would be inaccurate.) If we think the likelihood of server continuing to return UNSPECIFIED and changing future default is low enough, we can remove UNDEFINED from type and translate?

Reply:
We have to keep UNSPECIFIED in the protos, the enum needs this value to denote "unset"

Reply:
Loren: When do you want to unset? Do you mean an update where I want to set policy to current&future server defaults?

Roey: I mean indicating in gRPC that the value has not been set (undefined).

---

I didn't mean to suggest it be taken out of the `.proto`. What I was talking about here:

> likelihood of server continuing to return UNSPECIFIED

server will never return UNSPECIFIED from a Describe if this is implemented https://github.com/temporalio/temporal/issues/3240